File: scratch_fix_inventory.js
const fs = require('fs');
const path = './src/pages/inventory/InventoryLayout.jsx';

let content = fs.readFileSync(path, 'utf8');

// 1. In handleSubmitTransaction, replace:
// newTransactions.forEach(t => updateInventory(t));
// with:
// for (const t of newTransactions) { await updateInventory(t); }

content = content.replace(
  'newTransactions.forEach(t => updateInventory(t));',
  'for (const t of newTransactions) { await updateInventory(t); }'
);

// 2. In handleExcelDataSubmit, replace:
// inventoryUpdates.forEach(transaction => { updateInventory(transaction); });
// with:
// for (const transaction of inventoryUpdates) { await updateInventory(transaction); }

content = content.replace(
  /inventoryUpdates\.forEach\(transaction => \{\s*updateInventory\(transaction\);\s*\}\);/,
  'for (const transaction of inventoryUpdates) { await updateInventory(transaction); }'
);

// 3. Rewrite updateInventory completely
// Extract the current updateInventory implementation
const oldUpdateInventoryStr = `  const updateInventory = async (transaction) => {
    setInventory(prev => {
      const newInventory = { ...prev };
      
      if (transaction.type === 'in') {
        // 입고 처리 - 목적지가 창고인 경우만 재고 증가
        if (warehouses.find(w => w.id === transaction.toLocation)) {
          if (!newInventory[transaction.toLocation]) {
            newInventory[transaction.toLocation] = {};
          }
          if (!newInventory[transaction.toLocation][transaction.productId]) {
            newInventory[transaction.toLocation][transaction.productId] = 0;
          }
          newInventory[transaction.toLocation][transaction.productId] += transaction.quantity;
          
          // 서버에 재고 업데이트
          inventoryApi.upsert(transaction.toLocation, transaction.productId, newInventory[transaction.toLocation][transaction.productId]);
          
          // 창고 A(연동 창고)의 경우 실제 상품 재고도 업데이트
          const warehouse = warehouses.find(w => w.id === transaction.toLocation);
          if (warehouse?.syncWithProductStock) {
            updateProductStock(transaction.productId, transaction.quantity, 'increase');
          }
        }
        
        // 출발지가 창고/대리점인 경우 (창고→창고, 대리점→창고 이동) 출발지에서 재고 차감
        if (warehouses.find(w => w.id === transaction.fromLocation)) {
          if (newInventory[transaction.fromLocation] && 
              newInventory[transaction.fromLocation][transaction.productId]) {
            newInventory[transaction.fromLocation][transaction.productId] -= transaction.quantity;
            
            // 서버에 재고 업데이트
            inventoryApi.upsert(transaction.fromLocation, transaction.productId, newInventory[transaction.fromLocation][transaction.productId]);
            
            // 창고 A(연동 창고)의 경우 실제 상품 재고도 업데이트
            const warehouse = warehouses.find(w => w.id === transaction.fromLocation);
            if (warehouse?.syncWithProductStock) {
              updateProductStock(transaction.productId, -transaction.quantity, 'decrease');
            }
          }
        }
      } else {
        // 출고 처리 - 출발지가 창고인 경우만 재고 차감
        if (warehouses.find(w => w.id === transaction.fromLocation)) {
          if (newInventory[transaction.fromLocation] && 
              newInventory[transaction.fromLocation][transaction.productId]) {
            newInventory[transaction.fromLocation][transaction.productId] -= transaction.quantity;
            
            // 서버에 재고 업데이트
            inventoryApi.upsert(transaction.fromLocation, transaction.productId, newInventory[transaction.fromLocation][transaction.productId]);
            
            // 창고 A(연동 창고)의 경우 실제 상품 재고도 업데이트
            const warehouse = warehouses.find(w => w.id === transaction.fromLocation);
            if (warehouse?.syncWithProductStock) {
              updateProductStock(transaction.productId, -transaction.quantity, 'decrease');
            }
          }
        }
        
        // 목적지가 창고인 경우만 재고 증가 (창고→창고 이동)
        if (warehouses.find(w => w.id === transaction.toLocation)) {
          if (!newInventory[transaction.toLocation]) {
            newInventory[transaction.toLocation] = {};
          }
          if (!newInventory[transaction.toLocation][transaction.productId]) {
            newInventory[transaction.toLocation][transaction.productId] = 0;
          }
          newInventory[transaction.toLocation][transaction.productId] += transaction.quantity;
          
          // 서버에 재고 업데이트
          inventoryApi.upsert(transaction.toLocation, transaction.productId, newInventory[transaction.toLocation][transaction.productId]);
          
          // 창고 A(연동 창고)의 경우 실제 상품 재고도 업데이트
          const warehouse = warehouses.find(w => w.id === transaction.toLocation);
          if (warehouse?.syncWithProductStock) {
            updateProductStock(transaction.productId, transaction.quantity, 'increase');
          }
        }
        
        // 목적지가 대리점인 경우는 재고 관리하지 않음 (출고 기록만)
      }
      
      return newInventory;
    });
  };`;

const newUpdateInventoryStr = `  const updateInventory = async (transaction) => {
    let nextInventory = null;
    
    // 1. 로컬 상태 계산
    await new Promise(resolve => {
      setInventory(prev => {
        const newInventory = JSON.parse(JSON.stringify(prev || {}));
        
        if (transaction.type === 'in') {
          if (warehouses.find(w => w.id === transaction.toLocation)) {
            if (!newInventory[transaction.toLocation]) newInventory[transaction.toLocation] = {};
            if (!newInventory[transaction.toLocation][transaction.productId]) newInventory[transaction.toLocation][transaction.productId] = 0;
            newInventory[transaction.toLocation][transaction.productId] += transaction.quantity;
          }
          if (warehouses.find(w => w.id === transaction.fromLocation)) {
            if (newInventory[transaction.fromLocation] && newInventory[transaction.fromLocation][transaction.productId]) {
              newInventory[transaction.fromLocation][transaction.productId] -= transaction.quantity;
            }
          }
        } else {
          if (warehouses.find(w => w.id === transaction.fromLocation)) {
            if (newInventory[transaction.fromLocation] && newInventory[transaction.fromLocation][transaction.productId]) {
              newInventory[transaction.fromLocation][transaction.productId] -= transaction.quantity;
            }
          }
          if (warehouses.find(w => w.id === transaction.toLocation)) {
            if (!newInventory[transaction.toLocation]) newInventory[transaction.toLocation] = {};
            if (!newInventory[transaction.toLocation][transaction.productId]) newInventory[transaction.toLocation][transaction.productId] = 0;
            newInventory[transaction.toLocation][transaction.productId] += transaction.quantity;
          }
        }
        
        nextInventory = newInventory;
        resolve();
        return newInventory;
      });
    });

    // 2. 서버 재고 업데이트 및 동기화 (await 적용)
    if (transaction.type === 'in') {
      if (warehouses.find(w => w.id === transaction.toLocation)) {
        await inventoryApi.upsert(transaction.toLocation, transaction.productId, nextInventory[transaction.toLocation][transaction.productId]);
        const warehouse = warehouses.find(w => w.id === transaction.toLocation);
        if (warehouse?.syncWithProductStock) {
          await updateProductStock(transaction.productId, transaction.quantity, 'increase');
        }
      }
      if (warehouses.find(w => w.id === transaction.fromLocation)) {
        await inventoryApi.upsert(transaction.fromLocation, transaction.productId, nextInventory[transaction.fromLocation][transaction.productId]);
        const warehouse = warehouses.find(w => w.id === transaction.fromLocation);
        if (warehouse?.syncWithProductStock) {
          await updateProductStock(transaction.productId, -transaction.quantity, 'decrease');
        }
      }
    } else {
      if (warehouses.find(w => w.id === transaction.fromLocation)) {
        await inventoryApi.upsert(transaction.fromLocation, transaction.productId, nextInventory[transaction.fromLocation][transaction.productId]);
        const warehouse = warehouses.find(w => w.id === transaction.fromLocation);
        if (warehouse?.syncWithProductStock) {
          await updateProductStock(transaction.productId, -transaction.quantity, 'decrease');
        }
      }
      if (warehouses.find(w => w.id === transaction.toLocation)) {
        await inventoryApi.upsert(transaction.toLocation, transaction.productId, nextInventory[transaction.toLocation][transaction.productId]);
        const warehouse = warehouses.find(w => w.id === transaction.toLocation);
        if (warehouse?.syncWithProductStock) {
          await updateProductStock(transaction.productId, transaction.quantity, 'increase');
        }
      }
    }
  };`;

// replace old with new
content = content.replace(oldUpdateInventoryStr, newUpdateInventoryStr);

fs.writeFileSync(path, content, 'utf8');
console.log('Fixed InventoryLayout.jsx async updates');

File: scratch_fix_inventory.py
import re
import sys

file_path = '/Users/weird/Desktop/Work/AS/crm-app/src/pages/inventory/InventoryLayout.jsx'
with open(file_path, 'r', encoding='utf-8') as f:
    content = f.read()

# 1. Update parseInt
# We replace parseInt(X) with parseInt(X, 10)
# We only match simple cases that don't have commas inside them
# For nested structures, we'll replace specifically

replacements = [
    ("parseInt(item.productId)", "parseInt(item.productId, 10)"),
    ("parseInt(item.quantity)", "parseInt(item.quantity, 10)"),
    ("parseInt(rowData[product.col])", "parseInt(rowData[product.col], 10)"),
    ("parseInt(rowData[4])", "parseInt(rowData[4], 10)"),
    ("parseInt(product.id)", "parseInt(product.id, 10)"),
    ("parseInt(it.quantity)", "parseInt(it.quantity, 10)"),
    ("parseInt(g.items[0].quantity)", "parseInt(g.items[0].quantity, 10)"),
    ("parseInt(product.productId)", "parseInt(product.productId, 10)"),
    ("parseInt(product.quantity)", "parseInt(product.quantity, 10)"),
    ("parseInt(e.target.value)", "parseInt(e.target.value, 10)")
]

for old, new in replacements:
    content = content.replace(old, new)

# 2. Update updateInventory function
old_update = """  const updateInventory = async (transaction) => {
    let nextInventory = null;
    
    // 1. 로컬 상태 계산
    await new Promise(resolve => {
      setInventory(prev => {
        const newInventory = JSON.parse(JSON.stringify(prev || {}));
        
        if (transaction.type === 'in') {
          if (warehouses.find(w => w.id === transaction.toLocation)) {
            if (!newInventory[transaction.toLocation]) newInventory[transaction.toLocation] = {};
            if (!newInventory[transaction.toLocation][transaction.productId]) newInventory[transaction.toLocation][transaction.productId] = 0;
            newInventory[transaction.toLocation][transaction.productId] += transaction.quantity;
          }
          if (warehouses.find(w => w.id === transaction.fromLocation)) {
            if (newInventory[transaction.fromLocation] && newInventory[transaction.fromLocation][transaction.productId]) {
              newInventory[transaction.fromLocation][transaction.productId] -= transaction.quantity;
            }
          }
        } else {
          if (warehouses.find(w => w.id === transaction.fromLocation)) {
            if (newInventory[transaction.fromLocation] && newInventory[transaction.fromLocation][transaction.productId]) {
              newInventory[transaction.fromLocation][transaction.productId] -= transaction.quantity;
            }
          }
          if (warehouses.find(w => w.id === transaction.toLocation)) {
            if (!newInventory[transaction.toLocation]) newInventory[transaction.toLocation] = {};
            if (!newInventory[transaction.toLocation][transaction.productId]) newInventory[transaction.toLocation][transaction.productId] = 0;
            newInventory[transaction.toLocation][transaction.productId] += transaction.quantity;
          }
        }
        
        nextInventory = newInventory;
        resolve();
        return newInventory;
      });
    });"""

new_update = """  const updateInventory = async (transaction) => {
    let nextInventory = null;
    let inventoryError = null;
    
    // 1. 로컬 상태 계산
    await new Promise(resolve => {
      setInventory(prev => {
        const newInventory = JSON.parse(JSON.stringify(prev || {}));
        
        try {
          if (transaction.type === 'in') {
            if (warehouses.find(w => w.id === transaction.toLocation)) {
              if (!newInventory[transaction.toLocation]) newInventory[transaction.toLocation] = {};
              if (!newInventory[transaction.toLocation][transaction.productId]) newInventory[transaction.toLocation][transaction.productId] = 0;
              newInventory[transaction.toLocation][transaction.productId] += transaction.quantity;
            }
            if (warehouses.find(w => w.id === transaction.fromLocation)) {
              if (newInventory[transaction.fromLocation] && newInventory[transaction.fromLocation][transaction.productId]) {
                const current = newInventory[transaction.fromLocation][transaction.productId];
                if (current < transaction.quantity) throw new Error(`재고 부족: [${transaction.fromLocation}]에서 [${transaction.productName}]의 재고가 부족합니다. (현재: ${current})`);
                newInventory[transaction.fromLocation][transaction.productId] -= transaction.quantity;
              }
            }
          } else {
            if (warehouses.find(w => w.id === transaction.fromLocation)) {
              if (newInventory[transaction.fromLocation] && newInventory[transaction.fromLocation][transaction.productId]) {
                const current = newInventory[transaction.fromLocation][transaction.productId];
                if (current < transaction.quantity) throw new Error(`재고 부족: [${transaction.fromLocation}]에서 [${transaction.productName}]의 재고가 부족합니다. (현재: ${current})`);
                newInventory[transaction.fromLocation][transaction.productId] -= transaction.quantity;
              }
            }
            if (warehouses.find(w => w.id === transaction.toLocation)) {
              if (!newInventory[transaction.toLocation]) newInventory[transaction.toLocation] = {};
              if (!newInventory[transaction.toLocation][transaction.productId]) newInventory[transaction.toLocation][transaction.productId] = 0;
              newInventory[transaction.toLocation][transaction.productId] += transaction.quantity;
            }
          }
        } catch (e) {
          inventoryError = e;
          resolve();
          return prev;
        }
        
        nextInventory = newInventory;
        resolve();
        return newInventory;
      });
    });

    if (inventoryError) throw inventoryError;"""

content = content.replace(old_update, new_update)

with open(file_path, 'w', encoding='utf-8') as f:
    f.write(content)
print("Updated successfully")
